Add docker compose setup

philipgatzka opened the pull request:
Hi there 👋

I wanted to give this project a try and have gone ahead and added a simple docker-compose setup.

`make up` executes all necessary setup tasks as described in the `README.md` and runs the application on `localhost:4000`. The only requirements to use this are `make`, `docker` and `docker compose`.

This might be a solution to #50 . I'm open for suggestions and changes 😊

philipgatzka commented on php-cli.Dockerfile:
```diff
@@ -0,0 +1,30 @@
+
+RUN <<EOF
+install-php-extensions \
+    apcu \
+    bcmath \
+    bz2 \
+    calendar \
+    ffi \
+    gd \
+    gmp \
+    imagick \
+    intl \
+    mysqli \
+    pcntl \
+    pcov \
+    pdo_mysql \
+    pdo_pgsql \
+    redis \
+    soap \
+    sockets \
+    sodium \
+    xsl \
+    zip \
+    exif
```
I'm not sure, whether all of these are actually needed, so it might be possible to slim the image down and shorten the build duration a bit.

JhumanJ reviewed (commented):
That's awesome, thank you! I'll give this a try asap!

JhumanJ commented:
Hey @philipgatzka could you please share more information on how to use this? I'm not very used to docker. 
I tried 'docker compose up' and `make` and I it failed both times at the migration step

```
jwt-auth secret [sgCYVZn2bsiY5GG93rh6iLR1qfgzv35UGLMhHsvmfM9o1LRgg3uXaEFo2WcU8jXH] set successfully.
touch .make.jwt-secret
docker compose run --rm migrate
[+] Running 1/0
 ⠿ Container opnform-database-1  Created                                                                                                                                        0.1s
[+] Running 1/1
 ⠿ Container opnform-database-1  Started                                                                                                                                        6.8s

   Illuminate\Database\QueryException 

  SQLSTATE[08006] [7] could not connect to server: Connection refused
        Is the server running on host "127.0.0.1" and accepting
        TCP/IP connections on port 5432? (SQL: select * from information_schema.tables where table_catalog = openform and table_schema = public and table_name = migrations and table_type = 'BASE TABLE')

  at vendor/laravel/framework/src/Illuminate/Database/Connection.php:760
    756▕         // If an exception occurs when attempting to run a query, we'll format the error
    757▕         // message to include the bindings with SQL, which will make this exception a
    758▕         // lot more helpful to the developer instead of just the database's errors.
    759▕         catch (Exception $e) {
  ➜ 760▕             throw new QueryException(
    761▕                 $query, $this->prepareBindings($bindings), $e
    762▕             );
    763▕         }
    764▕     }

      +36 vendor frames 
  37  artisan:37
      Illuminate\Foundation\Console\Kernel::handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))make: *** [.make.migrate] Error 1
```

philipgatzka commented:
Hi @JhumanJ , the PHP container can't find the database. I assume this is due to an already existing `.env` file.

`make up` copies the `.env.dist` to `.env`, substituting the database server address of `127.0.0.1` with the database container's hostname `database`. This substitution only takes place, though, when the file doesn't exist already.

Please try to substitute `DB_HOST=127.0.0.1` with `DB_HOST=database` and let me know how it goes 😊

I'm on mobile currently and will add some documentation next time I get to work on this.

JhumanJ reviewed (changes requested):
Hey @philipgatzka , apologies for the late reply. I just gave this another shot:
- When running `make up` the first time it fails:
```
$make up
-----
docker compose down
[+] Running 2/1
 ⠿ Container opnform-database-1  Removed                                                                                                                                        0.1s
 ⠿ Network opnform_default       Removed                                                                                                                                        0.1s
cp .env.example .env
sed -i 's/DB_HOST=127.0.0.1/DB_HOST=database/' .env
sed: 1: ".env": invalid command code .
make: *** [.env] Error 1
```
The second time it works well.

- I'm getting some database errors:
```
⠿ Container opnform-database-1  Started                                                                                                                                        0.2s

   Illuminate\Database\QueryException 

  SQLSTATE[08006] [7] could not connect to server: Connection refused
        Is the server running on host "127.0.0.1" and accepting
        TCP/IP connections on port 5432? (SQL: select * from information_schema.tables where table_catalog = postgres and table_schema = public and table_name = migrations and table_type = 'BASE TABLE')

  at vendor/laravel/framework/src/Illuminate/Database/Connection.php:760
    756▕         // If an exception occurs when attempting to run a query, we'll format the error
    757▕         // message to include the bindings with SQL, which will make this exception a
    758▕         // lot more helpful to the developer instead of just the database's errors.
    759▕         catch (Exception $e) {
  ➜ 760▕             throw new QueryException(
    761▕                 $query, $this->prepareBindings($bindings), $e
    762▕             );
    763▕         }
    764▕     }

      +36 vendor frames 
  37  artisan:37
      Illuminate\Foundation\Console\Kernel::handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
make: *** [.make.migrate] Error 1
```

Would you have some time to take a look at this? I'd be more than happy to do some pair coding on this issue. I think we're almost there and getting this docker setup would really help others getting started with the project.
Thanks again for your help 🙏

philipgatzka commented:
Hi @JhumanJ , I've added a few sentences to the readme (please feel free to adjust to your liking) and also removed the dependency on `sed`, since it behaves differently on Linux and OSX systems. The compose setup now has it's own `.env.docker` file, which contains the database connection information that differs from the local setup. This way, the existing `.env` doesn't need to be altered at all and one should be able to switch between local and docker setup without needing to fiddle with config files.

I hope this will work more robustly now 😄

> I'd be more than happy to do some pair coding on this issue.

Great, I can't promise I'll be readily available, but in case you'd like to iron out more bugs, let me know how/when to contact you and I'll try to make time.

JhumanJ commented:
It works perfectly, thank you so much @philipgatzka 🙏 

If you have some more time here are things that I think should be added/change:
- Ngninx instead of `php artisan serve`
- Add a [Minio](https://min.io/) image to replace the need for S3
- Add a task to run the queue

Again, happy to collaborate with you on these if you're interested!

philipgatzka commented:
Glad to hear! Also great ideas for improvement 👍
